README.md
# OpenShift Extension for Visual Studio Code

This extension has been removed in favor of the official extension developed by Red Hat!

Check it out here:

https://marketplace.visualstudio.com/items?itemName=redhat.vscode-openshift-connector

Thanks!

----
If you need my old plugin code for any reason, that can still be [found on github here](https://github.com/dudash/openshift-vscode-extension)

CHANGELOG.md
# Change Log
All notable changes to the "openshift-vscode-extension" extension will be documented in this file.

## [Final - Deprecated]
- This plugin has been removed in favor of [the official openshift plugin from Red Hat](https://marketplace.visualstudio.com/items?itemName=redhat.vscode-openshift-connector)

## [Prerelease]
- Not tracking changes until v1.0

apiutils.js
// ----------------------------------------------------------------------------
// Author: Jason Dudash
// Born on: 2017-04-28
// !!!!!!!!!!!!!!!!!!!!!!!!!!!
// WARNING: this is a weekend prototype extension for OpenShift, 
// the full version should organize the code into multiple files with 
// a smarter design, std logging, better aync call design, etc... 
// !!!!!!!!!!!!!!!!!!!!!!!!!!!
// ----------------------------------------------------------------------------

var fetch = require('node-fetch');

// need this for a custom https agent, see here: https://github.com/bitinn/node-fetch/issues/15
var https = require('https');
const agent = new https.Agent({ rejectUnauthorized: false });

// yes, pretty hacky - but don't want to pass from extension.js everytime
// it will get updated on login calls only
// probably will write some sort of session class longer term...
var token = 'FLAG_NEED_TOKEN';
var masterURL = 'FLAG_NEED_MASTERURL';
var insecure = true;

const OAPIPROJ = "/oapi/v1/projects";
const APINS_PREFIX = "/api/v1/namespaces";
const OAPINS_PREFIX = "/oapi/v1/namespaces";
const PATH_PODS = "/pods";
const PATH_BUILDS = "/builds";
const PATH_BCS = "/buildconfigs";

//--------------------------------------------------------------------------------
var oapiLogin = function(m, t, i) {
    masterURL = m;
    var pattern = /^((http|https|ftp):\/\/)/;
    if(!pattern.test(masterURL)) {
        masterURL = "https://" + masterURL;
    }
    // TOOD: test for port on end and append 8443 if missing?
    
    token = t;
    
    insecure = i;
    // TODO: update agent with insecure setting

    return new Promise((resolve, reject) => {
        // TODO: in the future login with user/pass
        resolve();
    });
}

//--------------------------------------------------------------------------------
var oapiLogout = function() {
    return new Promise((resolve, reject) => {
        // TODO:
        resolve();
    });
}

//--------------------------------------------------------------------------------
var oapiGetProjects = function() {
    return new Promise((resolve, reject) => {
        var url = masterURL + OAPIPROJ;
        var opts = {
            method: 'GET',
            headers: { 'Authorization': 'Bearer ' + token },
			agent: agent
		};
        console.log('GET from ' + url + ' with token ' + token + ', insecure=' + insecure + '\n');
        fetch(url, opts)
            .then(function(res) {
                console.log('got a response from ' + url + '\n');
                return res.json();
            }).then(function(json) {
                console.log(json);
                console.log('------');
                console.log(convertProjectsJsonToArray(json));
                resolve(json);
            }).catch(function(err) {
                console.log(err);
                reject(err);
            });
    });
}

//--------------------------------------------------------------------------------
var oapiSetProject = function(project) {
    return new Promise((resolve, reject) => {
        var url = masterURL + OAPIPROJ + project;
        var opts = {
            method: 'GET',
            headers: { 'Authorization': 'Bearer ' + token },
			agent: agent
		};
        console.log('GET from ' + url + ' with token ' + token + ', insecure=' + insecure + '\n');
        fetch(url, opts)
            .then(function(res) {
                console.log('got a response from ' + url + '\n');
                return res.json();
            }).then(function(json) {
                console.log(json);
                resolve(json);
            }).catch(function(err) {
                console.log(err);
                reject(err);
            });
    });
}

//--------------------------------------------------------------------------------
function convertProjectsJsonToArray(json) {
    var items = json.items;
    var projects = [];
    for(var i = 0; i < items.length; i++)
    {
        var name = items[i].metadata.name;
        projects.push(name);
    }
    return projects;
}

//--------------------------------------------------------------------------------
var oapiDescribe = function(item, inproject) {
    return new Promise((resolve, reject) => {
        var url = masterURL + OAPINS_PREFIX + inproject + '/' + item;
        var opts = {
            method: 'GET',
            headers: { 'Authorization': 'Bearer ' + token },
			agent: agent
		};
        console.log('GET from ' + url + ' with token ' + token + ', insecure=' + insecure + '\n');
        fetch(url, opts)
            .then(function(res) {
                console.log('got a response from ' + url + '\n');
                return res.text();
            }).then(function(body) {
                console.log(body);
                resolve(body);
            }).catch(function(err) {
                console.log(err);
                reject(err);
            });
    });
}

//--------------------------------------------------------------------------------
var oapiStatus = function(inproject) {
    return new Promise((resolve, reject) => {
        var url = masterURL + OAPINS_PREFIX + inproject ;
        var opts = {
            method: 'GET',
            headers: { 'Authorization': 'Bearer ' + token },
			agent: agent
		};
        // TODO: pull a bunch of status and consolidate into an single output
        resolve();
    });
}

//--------------------------------------------------------------------------------
var oapiPods = function(inproject) {
    return new Promise((resolve, reject) => {
        var url = masterURL + APINS_PREFIX + inproject + PATH_PODS;
        var opts = {
            method: 'GET',
            headers: { 'Authorization': 'Bearer ' + token },
			agent: agent
		};
        console.log('GET from ' + url + ' with token ' + token + ', insecure=' + insecure + '\n');
        fetch(url, opts)
            .then(function(res) {
                console.log('got a response from ' + url + '\n');
                return res.json();
            }).then(function(json) {
                console.log(json);
                resolve(json);
            }).catch(function(err) {
                console.log(err);
                reject(err);
            });
    });
}

//--------------------------------------------------------------------------------
var oapiBuilds = function(inproject) {
    return new Promise((resolve, reject) => {
        var url = masterURL + OAPINS_PREFIX + inproject + PATH_BUILDS;
        var opts = {
            method: 'GET',
            headers: { 'Authorization': 'Bearer ' + token },
			agent: agent
		};
        console.log('GET from ' + url + ' with token ' + token + ', insecure=' + insecure + '\n');
        fetch(url, opts)
            .then(function(res) {
                console.log('got a response from ' + url + '\n');
                return res.json();
            }).then(function(json) {
                console.log(json);
                resolve(json);
            }).catch(function(err) {
                console.log(err);
                reject(err);
            });
    });
}

//--------------------------------------------------------------------------------
var oapiBuildConfigs = function(inproject) {
    return new Promise((resolve, reject) => {
        var url = masterURL + OAPINS_PREFIX + inproject + PATH_BCS;
        var opts = {
            method: 'GET',
            headers: { 'Authorization': 'Bearer ' + token },
			agent: agent
		};
        console.log('GET from ' + url + ' with token ' + token + ', insecure=' + insecure + '\n');
        fetch(url, opts)
            .then(function(res) {
                console.log('got a response from ' + url + '\n');
                return res.json();
            }).then(function(json) {
                console.log(json);
                console.log('------');
                console.log(convertBuildConfigsJsonToArray(json));
                resolve(json);
            }).catch(function(err) {
                console.log(err);
                reject(err);
            });
    });
}

//--------------------------------------------------------------------------------
function convertBuildConfigsJsonToArray(json) {
    var items = json.items;
    var bcArray = [];
    for(var i = 0; i < items.length; i++)
    {
        var name = items[i].metadata.name;
        bcArray.push(name);
    }
    return bcArray;
}

//--------------------------------------------------------------------------------
var oapiLogs = function(inproject, resource) {
    return new Promise((resolve, reject) => {
        var url = masterURL + OAPINS_PREFIX + inproject + PATH_BCS + '/' + resource;
        var opts = {
            method: 'GET',
            headers: { 'Authorization': 'Bearer ' + token },
			agent: agent
		};
        console.log('GET from ' + url + ' with token ' + token + ', insecure=' + insecure + '\n');
        fetch(url, opts)
            .then(function(res) {
                console.log('got a response from ' + url + '\n');
                return res.text();
            }).then(function(body) {
                console.log(body);
                resolve(body);
            }).catch(function(err) {
                console.log(err);
                reject(err);
            });
    });
}

// export yo'shit
exports.oapiLogin = oapiLogin;
exports.oapiLogout = oapiLogout;
exports.oapiGetProjects = oapiGetProjects;
exports.oapiSetProject = oapiSetProject;
exports.convertProjectsJsonToArray = convertProjectsJsonToArray;
exports.oapiDescribe = oapiDescribe;
exports.oapiStatus = oapiStatus;
exports.oapiPods = oapiPods;
exports.oapiBuilds = oapiBuilds;
exports.oapiBuildConfigs = oapiBuildConfigs;
exports.convertBuildConfigsJsonToArray = convertBuildConfigsJsonToArray;
exports.oapiLogs = oapiLogs;


extension.js
// ----------------------------------------------------------------------------
// Author: Jason Dudash
// Born on: 2017-04-28
// !!!!!!!!!!!!!!!!!!!!!!!!!!!
// WARNING: this is a weekend prototype extension for OpenShift, 
// the full version should organize the code into multiple files with 
// a smarter design, std logging, better aync call design, etc... 
// !!!!!!!!!!!!!!!!!!!!!!!!!!!
// ----------------------------------------------------------------------------

// FUTURE: this looks like a good extension to aspire to be as complete as:
// https://github.com/bradygaster/azure-tools-vscode

// NOTE: Use the console to output diagnostic information (console.log) and errors (console.error)

var vscode = require('vscode'); // VS Code extensibility API
var apiutils = require('./apiutils');
var prettyit = require('./prettythejson');

var token = 'FLAG_NEED_TOKEN';
var masterURL = 'FLAG_NEED_MASTERURL';
var currentProject = '/FLAG_NEED_PROJECT';
var projectsList = [];

var openshiftOutput = vscode.window.createOutputChannel('OpenShift Output');
openshiftOutput.show();
var statusBarItem;

//--------------------------------------------------------------------------------
function activate(context) {
    // NOTE: commands need to match the package.json file
    console.log('extension "openshift-vscode-extension"');

    var loggedIn = false;
    var configSettings = vscode.workspace.getConfiguration('openshift');
    currentProject = '/' + configSettings.get('defaultProject');

    statusBarItem = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right);
    statusBarItem.text = 'OpenShift-Login';
    statusBarItem.tooltip = 'Click to Login';
    statusBarItem.command = 'openshift.login';
    statusBarItem.show();  // FIXME: this doesn't show before login for some reason

    // ****** Login Command
    var disposable = vscode.commands.registerCommand('extension.login', function () {
        if (loggedIn) {
            openshiftOutput.append('You will be logged out of: ' + masterURL + '\n');
            // TODO: warn you will be logged out and get approval
        }
        token = configSettings.get('token');
        masterURL = configSettings.get('masterURL');
        var insecure = configSettings.get('insecure', true);
        vscode.window.showInputBox({prompt: 'Input a master URL (esc to use ' + masterURL + ')'})
            .then(value => {
                if (value == undefined) { // user hit escape, that's OK - use defaults
                } else if (value.length > 1) { masterURL = value; }
                vscode.window.showInputBox({prompt: 'Input a connection token (esc to use ' + token + ')'})
                .then(value => {
                    if (value == undefined) { // user hit escape, that's OK - use defaults
                    } else if (value.length > 1) { token = value; }
                    openshiftOutput.append('Attempting to login to: ' + masterURL + ' with token ' + token + '\n');
                    apiutils.oapiLogin(masterURL, token, insecure)
                        .then(value => {
                            // FIXME: make sure login succeeded
                            loggedIn = true;
                            openshiftOutput.append('Logged into: ' + masterURL + ' with token ' + token + '\n');
                            
                            // poll for projects and add them to a status bar item for selection
                            apiutils.oapiGetProjects()
                                .then(value => {
                                    projectsList = apiutils.convertProjectsJsonToArray(value);
                                    openshiftOutput.append('Available projects are:' + projectsList + '\n\n');
                                    statusBarItem.text = currentProject;
                                    statusBarItem.tooltip = 'OpenShift - Logged into ' + masterURL;;
                                    statusBarItem.command = 'extension.selectProject';
                                    statusBarItem.show();
                                })
                                .catch(value => {
                                    openshiftOutput.append('ERROR GETTING PROJECTS LIST: \n' + value + '\n\n');
                                    return;
                                });
                        })
                        .catch(value => {
                            vscode.window.showErrorMessage('Error logging in');
                            openshiftOutput.append('ERROR: \n' + value + '\n\n');
                            return;
                        });
                });
            });
    });
    context.subscriptions.push(disposable);

    // ****** Logout Command
    var disposable = vscode.commands.registerCommand('extension.logout', function () {
        if (loggedIn) { openshiftOutput.append('You will be logged out of: ' + masterURL + '\n'); }
        apiutils.oapiLogout()
            .then(value => {
                openshiftOutput.append('LOGGED OUT \n');
            })
            .catch(value => {
                vscode.window.showErrorMessage('Error logging out');
                openshiftOutput.append('ERROR: \n' + value + '\n\n');
                return;
            });
        loggedIn = false;
        statusBarItem.text = 'OpenShift-Login';
        statusBarItem.tooltip = 'Click to Login';
        statusBarItem.command = 'extension.login';
    });
    context.subscriptions.push(disposable);

    // ****** Get Status Command
    disposable = vscode.commands.registerCommand('extension.getStatus', function () {
        if (!loggedIn) { vscode.window.showErrorMessage('Please login to an OpenShift cluster'); return; }
        
        //!!!!!!
        vscode.window.showErrorMessage('Status command not implemented yet');
        
        apiutils.oapiStatus(currentProject)
            .then(value => {
                openshiftOutput.append('STATUS: \n' + value + '\n\n');
            })
            .catch(value => {
                vscode.window.showErrorMessage('Error getting status');
                openshiftOutput.append('ERROR: \n' + value + '\n\n');
                return;
            });
    });
    context.subscriptions.push(disposable);

    // ****** Describe This Command
    disposable = vscode.commands.registerCommand('extension.describeThis', function () {
        if (!loggedIn) { vscode.window.showErrorMessage('Please login to an OpenShift cluster'); return; }
        var editor = vscode.window.activeTextEditor;
        if (!editor) { // No open text editor
            vscode.window.showErrorMessage('Must select text within an open file');
            return; 
        }
        var selection = editor.selection;
        var text = editor.document.getText(selection);
        if (text.length < 1) { // No selection
            vscode.window.showErrorMessage('Please select some text in your active text editor.');
            return; 
        }
        vscode.window.showInformationMessage('Describe: ' + text.valueOf());
        apiutils.oapiDescribe(text.valueOf(), currentProject)
            .then(value => {
                openshiftOutput.append('DESCRIBE ' + text.valueOf() + ' IN ' + currentProject + ':\n' + value + '\n\n');
            })
            .catch(value => {
                vscode.window.showErrorMessage('Error describing selected text');
                openshiftOutput.append('ERROR: \n' + value + '\n\n');
                return;
            });
    });
    context.subscriptions.push(disposable);

    // ****** Get Pods Command
    disposable = vscode.commands.registerCommand('extension.getPods', function () {
        if (!loggedIn) { vscode.window.showErrorMessage('Please login to an OpenShift cluster'); return; }
        apiutils.oapiPods(currentProject)
            .then(value => {
                // TODO: open results in a new file
                openshiftOutput.append('PODS IN ' + currentProject + ':\n' + JSON.stringify(value) + '\n\n');
            })
            .catch(value => {
                vscode.window.showErrorMessage('Error getting pods');
                openshiftOutput.append('ERROR: \n' + value + '\n\n');
                return;
            });
    });
    context.subscriptions.push(disposable);

    // ****** Get Builds Command
    disposable = vscode.commands.registerCommand('extension.getBuilds', function () {
        if (!loggedIn) { vscode.window.showErrorMessage('Please login to an OpenShift cluster'); return; }
        apiutils.oapiBuilds(currentProject)
            .then(value => {
                openshiftOutput.append('BUILDS IN ' + currentProject + ':\n' + JSON.stringify(value) + '\n\n');
            })
            .catch(value => {
                vscode.window.showErrorMessage('Error getting builds');
                openshiftOutput.append('ERROR: \n' + value + '\n\n');
                return;
            });
    });
    context.subscriptions.push(disposable);

    // ****** Logs Command
    disposable = vscode.commands.registerCommand('extension.getLogs', function () {
        if (!loggedIn) { vscode.window.showErrorMessage('Please login to an OpenShift cluster'); return; }

        //!!!!!!
        // vscode.window.showErrorMessage('Logs command not implemented yet');
        apiutils.oapiBuildConfigs(currentProject)
            .then(bcjson => {
                var buildConfigs = apiutils.convertBuildConfigsJsonToArray(bcjson);
                openshiftOutput.append('Available bcs are: ' + buildConfigs + '\n\n');
                vscode.window.showQuickPick(buildConfigs)
                .then(pick => {
                    if (!pick) { return; } // nothing selected
                    var selected = '/' + pick;
                    // make sure we can access it
                    apiutils.oapiLogs(currentProject, selected)
                        .then(logs => {
                            // TODO: parse logs to readable format
                            openshiftOutput.append('LOGS IN ' + currentProject + ' for bc' + selected + ':\n' + logs + '\n\n');
                        })
                        .catch(logserr => {
                            vscode.window.showErrorMessage('Error getting logs');
                            openshiftOutput.append('ERROR: \n' + logserr + '\n\n');
                            return;
                        });
                });
            })
            .catch(bcerr => {
                vscode.window.showErrorMessage('Error getting logs');
                openshiftOutput.append('ERROR: \n' + bcerr + '\n\n');
                return;
            });
    });
    context.subscriptions.push(disposable);

    // ****** Select Project Command
    disposable = vscode.commands.registerCommand('extension.selectProject', function () {
        if (!loggedIn) { vscode.window.showErrorMessage('Please login to an OpenShift cluster'); return; }
        vscode.window.showQuickPick(projectsList)
        .then(value => {
            if (!value) { return; } // nothing selected
            var selected = '/' + value;
            // make sure we can access it
            apiutils.oapiSetProject(selected)
                .then(value => {
                    currentProject = selected;
                    openshiftOutput.append('SELECTED PROJECT: ' + currentProject + '\n\n');
                    statusBarItem.text = currentProject;
                    statusBarItem.tooltip = 'OpenShift - Logged into ' + masterURL;;
                    statusBarItem.command = 'extension.selectProject';
                    statusBarItem.show();
                })
                .catch(value => {
                    vscode.window.showErrorMessage('Error settting new project');
                    openshiftOutput.append('ERROR: \n' + value + '\n\n');
                    return;
                });
        });
    });
    context.subscriptions.push(disposable);
}
exports.activate = activate;

//--------------------------------------------------------------------------------
function deactivate() {
    // TODO: clean up anything necessary (maybe logout?)
}

exports.deactivate = deactivate;